includes/Products/Feed.php

```diff
@@ -179,6 +187,192 @@ public function schedule_feed_generation() {
+			WC_Facebookcommerce_Utils::log( 'Feed: retrieve_integration_feed_id(): feed_id = '.$feed_id.', created a new feed via Meta API.');
+			return $feed_id;
+		}
+
+		return '';
+	}
+
+	/**
+	 * Validates that provided feed ID still exists on the Meta side
+	 *
+	 * @param 		string $feed_id the feed ID
+	 * 
+	 * @return		bool true if the feed ID is valid
+	 * 
+	 * @internal
+	 */
+	private function validate_feed_exists($feed_id) {
+		$catalog_id = facebook_for_woocommerce()->get_integration()->get_product_catalog_id();
+		if ( '' === $catalog_id ) {
+			throw new Error( 'No catalog ID' );
+		}
+		
+		try {
+			$feed_nodes = facebook_for_woocommerce()->get_api()->read_feeds( $catalog_id )->data;
+		} catch ( Exception $e ) {
+			$message = sprintf( 'There was an error trying to get feed nodes for catalog: %s', $e->getMessage() );
+			WC_Facebookcommerce_Utils::log( $message );
+			return '';
+		}
+
+		foreach ( $feed_nodes as $feed ) {
+			if ($feed['id'] == $feed_id) {
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	/**
+	 * Queries existing feeds for the integration catalog and filters
+	 * the plugin integration feed ID
+	 * 
+	 * @return		string the integration feed ID
+	 * 
+	 * @internal
+	 */
+	private function query_and_filter_integration_feed_id() {
+		$catalog_id = facebook_for_woocommerce()->get_integration()->get_product_catalog_id();
+		if ( '' === $catalog_id ) {
+			throw new Error( 'No catalog ID' );
+		}
+		
+		try {
+			$feed_nodes = facebook_for_woocommerce()->get_api()->read_feeds( $catalog_id )->data;
+		} catch ( Exception $e ) {
+			$message = sprintf( 'There was an error trying to get feed nodes for catalog: %s', $e->getMessage() );
+			WC_Facebookcommerce_Utils::log( $message );
+			return '';
+		}
+
+		if ( !empty( $feed_nodes ) ) {
+			
+			try {
+				$catalog = facebook_for_woocommerce()->get_api()->get_catalog( $catalog_id );
+			} catch ( Exception $e ) {
+				$message = sprintf( 'There was an error trying to get a catalog: %s', $e->getMessage() );
+				WC_Facebookcommerce_Utils::log( $message );
+			}
+
+			/*
+				We need to detect which feed is the one that was created for Facebook for WooCommerce plugin usage.
+
+				We are detecting based on the name.
+				- Option 1. Plugin can create this feed name currently.
+				- Option 2 and 3. FBE creates a catalog with feed name '{catalog name} - Feed' or '{catalog name} – Feed' (short vs long dash)
+				- Option 4. Plugin used to create a feed name 'Initial product sync from WooCommerce. DO NOT DELETE.'
+			*/
+			foreach ( $feed_nodes as $feed ) {
+				try {
+					$feed_metadata = facebook_for_woocommerce()->get_api()->read_feed( $feed['id'] );
+				} catch ( Exception $e ) {
+					$message = sprintf( 'There was an error trying to get feed metadata: %s', $e->getMessage() );
+					WC_Facebookcommerce_Utils::log( $message );
+					continue;
+				}
+
+				$woo_feed_name_option_1 = self::FEED_NAME;
```

Comment:
suggest factor out to private function named something like `private function pattern_match_feed_names()`